contracts/ThreeHeapOrdering.sol

```diff
@@ -0,0 +1,306 @@
+
+        uint256 size = _heap.size;
+        uint256 accountsLength = _heap.accounts.length;
+
+        if (size < accountsLength) {
+            Account memory firstAccountNotSorted = _heap.accounts[size];
+
+            _heap.indexes[firstAccountNotSorted.id] = accountsLength;
+            _heap.accounts.push(firstAccountNotSorted);
+        } else _heap.accounts.push();
+
+        shiftUp(_heap, Account(_id, _value), size);
+        _heap.size = computeSize(size + 1, _maxSortedUsers);
+    }
+
+    /// @notice Decreases the amount of an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with a value greater than `_newValue`.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to decrease the amount.
+    /// @param _newValue The new value of the account.
+    function decrease(
+        HeapArray storage _heap,
+        address _id,
+        uint96 _newValue
+    ) private {
+        uint256 index = _heap.indexes[_id];
+
+        if (index < (_heap.size - 1) / 3) shiftDown(_heap, Account(_id, _newValue), index);
+        else _heap.accounts[index].value = _newValue;
+    }
+
+    /// @notice Increases the amount of an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with a smaller value than `_newValue`.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to increase the amount.
+    /// @param _newValue The new value of the account.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    function increase(
+        HeapArray storage _heap,
+        address _id,
+        uint96 _newValue,
+        uint256 _maxSortedUsers
+    ) private {
+        uint256 index = _heap.indexes[_id];
+        _heap.accounts[index].value = _newValue;
+        uint256 size = _heap.size;
+
+        if (index < size) shiftUp(_heap, Account(_id, _newValue), index);
+        else {
+            Account memory firstAccountNotSorted = _heap.accounts[size];
+            setAccount(_heap, index, firstAccountNotSorted);
+            shiftUp(_heap, Account(_id, _newValue), size);
+            _heap.size = computeSize(size + 1, _maxSortedUsers);
+        }
+    }
+
+    /// @notice Removes an account in the `_heap`.
+    /// @dev Only call when this function `_id` is in the `_heap` with value `_removedValue`.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to remove.
+    /// @param _removedValue The value of the account to remove.
+    function remove(
+        HeapArray storage _heap,
+        address _id,
+        uint96 _removedValue
+    ) private {
+        uint256 index = _heap.indexes[_id];
+        delete _heap.indexes[_id];
+        uint256 accountsLength = _heap.accounts.length;
+
+        if (_heap.size == accountsLength) _heap.size--;
+        if (index == accountsLength - 1) {
+            _heap.accounts.pop();
+            return;
+        }
```

Comment:
I think this is less likely to happen, perhaps removing it save gas in general.

Reply:
We can't remove it easily because we pop `accounts` after that and assume that `index` is still in the range of `acounts`